sled-agent/src/services.rs

```diff
@@ -5650,4 +5812,303 @@ mod test {
+                generation: Generation::new().next(),
+                zones: existing
+                    .iter()
+                    .enumerate()
+                    .map(|(i, (zone, _))| {
+                        let mut zone = zone.clone();
+                        zone.filesystem_pool = Some(some_zpools[i].clone());
+                        zone
+                    })
+                    .collect(),
+            };
+            let reconciled = reconcile_running_zones_with_new_request(
+                existing.iter_mut().map(|(z, p)| (z, &*p)),
+                new_request.clone(),
+                log,
+            )
+            .expect("reconciled successfully");
+            assert_eq!(reconciled.zones_to_be_removed, HashSet::new());
+            assert_eq!(reconciled.zones_to_be_added, HashSet::new());
+            assert_eq!(
+                existing.iter().map(|(z, _)| z.clone()).collect::<Vec<_>>(),
+                new_request.zones,
+            );
+        }
+
+        // Test 3: We have some zones; the new config changes `filesystem_pool`
+        // to match our runtime pools (i.e., the #7229 fix) but also changes
+        // something else in the config for the final zone; we should attempt to
+        // remove and re-add that final zone.
+        {
+            let mut existing = vec![
+                (
+                    make_omicron_zone_config(None),
+                    ZpoolOrRamdisk::Zpool(some_zpools[0].clone()),
+                ),
+                (
+                    make_omicron_zone_config(None),
+                    ZpoolOrRamdisk::Zpool(some_zpools[1].clone()),
+                ),
+                (
+                    make_omicron_zone_config(None),
+                    ZpoolOrRamdisk::Zpool(some_zpools[2].clone()),
+                ),
+            ];
+            let new_request = OmicronZonesConfig {
+                generation: Generation::new().next(),
+                zones: existing
+                    .iter()
+                    .enumerate()
+                    .map(|(i, (zone, _))| {
+                        let mut zone = zone.clone();
+                        zone.filesystem_pool = Some(some_zpools[i].clone());
+                        if i == 2 {
+                            zone.zone_type = OmicronZoneType::Oximeter {
+                                address: "[::1]:10000".parse().unwrap(),
+                            };
+                        }
+                        zone
+                    })
+                    .collect(),
+            };
+            let reconciled = reconcile_running_zones_with_new_request(
+                existing.iter_mut().map(|(z, p)| (z, &*p)),
+                new_request.clone(),
+                log,
+            )
+            .expect("reconciled successfully");
+            assert_eq!(
+                reconciled.zones_to_be_removed,
+                existing
+                    .iter()
+                    .skip(2)
```

Comment:
This is definitely a nitpick, but since you have the luxury of `existing` being a Vec, this might be more readable as:

```
HashSet::from([existing[2].0]);
```

And below - basically, just using slice indices instead of skipping / taking from iterators?

(This might just be me, but I find range notation easier to grok than skip/take... but I guess they're equivalent, so if you feel strongly, I can live with this as-is)

Reply:
Yeah, that's quite a lot more readable, thanks: 25bf35163